be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/LaoValidator.scala

```diff
@@ -1,140 +1,328 @@
+                  updateLao.last_modified,
+                  validationError(s"stale 'last_modified' timestamp (${updateLao.last_modified})")
+                ),
+                checkWitnesses(
+                  rpcMessage,
+                  updateLao.witnesses,
+                  validationError(dupWitnessError)
+                ),
+                checkId(
+                  rpcMessage,
+                  expectedHash,
+                  updateLao.id,
+                  validationError(unexpectedID + " " + updateLao.id.toString)
+                ),
+                checkChannelType(
+                  rpcMessage,
+                  ObjectType.LAO,
+                  channel,
+                  dbActorRef,
+                  validationError(s"trying to write an UpdateLao message on wrong type of channel $channel")
+                )
+              )
+
+            case Some(Failure(ex: DbActorNAckException)) => Right(PipelineError(ex.code, s"validateUpdateLao failed : ${ex.message}", rpcMessage.getId))
+            case reply                                   => Right(PipelineError(ErrorCodes.SERVER_ERROR.id, s"validateUpdateLao failed : unexpected DbActor reply '$reply'", rpcMessage.getId))
+          }
 
-            if (!validateTimestampStaleness(data.last_modified)) {
-              Right(validationError(s"stale 'last_modified' timestamp (${data.last_modified})"))
-            } else if (!validateWitnesses(data.witnesses)) {
-              Right(validationError("duplicate witnesses keys"))
-            } else if (expectedHash != data.id) {
-              Right(validationError("unexpected id"))
-            } else if (!validateChannelType(ObjectType.LAO, channel)) {
-              Right(validationError(s"trying to write an UpdateLao message on wrong type of channel $channel"))
-            } else {
-              Left(rpcMessage)
-            }
-          case Some(Failure(ex: DbActorNAckException)) => Right(PipelineError(ex.code, s"validateUpdateLao failed : ${ex.message}", rpcMessage.getId))
-          case reply                                   => Right(PipelineError(ErrorCodes.SERVER_ERROR.id, s"validateUpdateLao failed : unexpected DbActor reply '$reply'", rpcMessage.getId))
-        }
-
-      case _ => Right(validationErrorNoMessage(rpcMessage.id))
+        case _ => Right(validationErrorNoMessage(rpcMessage.id))
+
+      }
     }
   }
+
+  /** Check if all witnesses are distinct
+    *
+    * @param rpcMessage
+    *   rpc message to validate
+    * @param witnesses
+    *   witnesses to check
+    * @param error
+    *   the error to forward in case the witnesses are not all distinct
+    * @return
+    *   GraphMessage: passes the rpcMessages to Left if successful right with pipeline error
+    */
+  private def checkWitnesses(rpcMessage: JsonRpcRequest, witnesses: List[PublicKey], error: PipelineError): GraphMessage = {
+    if (validateWitnesses(witnesses))
+      Left(rpcMessage)
+    else
+      Right(error)
+  }
```

Comment:
This function could be moved into `MessageDataContentValidator.scala` file (same place as the function used inside `validateWitness`, since it can surely be reused in future validations functions.

Reply:
Done !